Update Python version support: drop 3.10 and add 3.14

DougBurke opened the pull request:
# Summary

Drop support for building with Python 3.10 and add support for building with Python 3.14.

# Details

At the moment there is only one change that we take advantage of, that we do not need to install the `tomli` package when building Sherpa. However, Python 3.10 is well into it's [only security fixes](https://devguide.python.org/versions/) phase, and we have historically only focused on supporting three versions. We could retain the 3.10 support but only test with 3.11 and above.

One of the CI runs is changed to build with Python 3.14 (one of the "minimal test" cases), to act as a canary ni case there are Python 3.14 issues, but so far it looks good for advertising 3.14 support.

The CI runs have been updated to use Python 3.11 to 3.14 but the choices made, in particular in the choice of versions of related code, such as NumPy, could do with a review.

There are no actual changes to code to add Python 3.11-only features in this PR.

DougBurke committed:
CI: bump python versions to 3.11/12/13

Drop Python 3.10 and add 3.13 from the CI runs.

As part of this, drop testing with NumPy version 1. It is not clear
what NumPy versions to use so either force version 2 or use a version
we are currently testing CIAO development code with.

Another change is that the CI runs now just install python rather
than force a cpython variant, as this was causing problems with the
testing of Python 3.13.

DougBurke committed:
BUILD: drop Python 3.10 support

This lets us not require the tomli package for building Sherpa.

hamogu commented:
I'm in favor of fully dropping 3.10, not just say "we don't test against it any more". We've been working on adding typing (which has successfully uncovered bugs and problems in the past) and the typing system is still improving from Python version to Python version. For example in 3.11, we can used "typing.Self", e.g. for class methods that generate a instance of the class (https://github.com/sherpa/sherpa/pull/2355/files#diff-39d64572ec8588c14cccff43dc2d85801af7524041351c8434f80963e847204bR5438).

DougBurke committed:
Note that Sherpa can be built with Python 3.14

There has been limited testing with Python 3.14, but enough to claim
we can build with it.

DougBurke committed:
CI: bump a conda run to Python 3.14

Provide a "minimal test" environment with Python 3.14 (replacing the
Python 3.12 version).

DougBurke committed:
Update README to point out Python 3.11 to 3.13 support.

DougBurke commented:
Rebasing.

DougBurke commented:
Rebasing to address conflicts and because #2457 was merged.
